python_modules/dagster/dagster/_daemon/sensor.py

```diff
@@ -875,6 +887,31 @@ def submit_run_request(run_request) -> SubmitRunRequestResult:
     yield
 
 
+def _handle_backfill_requests(
+    run_requests: Sequence[RunRequest],
+    instance: DagsterInstance,
+    context: SensorLaunchContext,
+) -> None:
+    for run_request in run_requests:
+        # TODO handling of duplicate run_keys so the same backfill isn't submitted twice
```

Comment:
i think this is pretty important, particularly since duplicate backfills are potentially much more disruptive than runs if they retry multiple times

one thing that we do in both asset backfills and AMP now in order to improve idempotency if a tick fails partway through is to reserve a run ID first and write it to the tick, so that if it fails partway through we can pick back up where we left off: https://github.com/dagster-io/dagster/blob/eea8da90ea79f77997340fdee539d756baa27449/python_modules/dagster/dagster/_daemon/asset_daemon.py#L927-L931

We probably should start doing that in the sensor daemon too, potentially for both run IDs and backfill IDs - maybe this is an excuse to do that? So we could reserve a backfill ID here, write it to the tick along with the run requests, and check for it if a tick retries?

Reply:
I think that, in most cases, it doesn't make to set a `run_key` when requesting asset executions. My understanding is that a `run_key` typically corresponds to a single "impetus" to launch a run.  E.g. "a new file landed in this directory so let's kick off a run of the job that should run once per file in that directory". But, when requesting assets, there's often multiple impetuses for a single run. E.g. "we're requesting this asset because its upstream asset has new data, and we're also requesting this other asset because the other asset's upstream asset has new data".

Reply:
> We probably should start doing that in the sensor daemon too, potentially for both run IDs and backfill IDs - maybe this is an excuse to do that? So we could reserve a backfill ID here, write it to the tick along with the run requests, and check for it if a tick retries?

(This scheme makes sense to me though)

Reply:
my understanding was that `run_key` served a slightly different purpose than recovering from tick failure. I see how reserving the run/backfill ids can help recover from tick failure, and i think that's a good idea to implement here. But I don't get how it solves the other use case for `run_key` of "I already requested the asset for this file N ticks ago, so don't run it again". 

I also get the point that `run_key` doesn't always make sense for assets, but it would be behavior change to stop supporting `run_key` for assets, and also strange if `run_key` works when you request a run using `asset_selection`, but doesn't work if you request a backfill.

Reply:
with the new method for constructing a `RunRequest` from an asset graph subset you can't use `run_key` so we don't need to consider it here anymore. In a future iteration we can consider adding `run_key` support and how that would be handled for backfills